Login Campaign

kaincenteno opened the pull request:
Adds logic for adding points if campaign is active.
Makes Greeter Moogles show list of items to trade points for.
Shows messages when zoning/logging in about Campaign.

Only three past campaigns were added, defaulting to December's but it won't run
as that campaign was only meant to run for 23 days.

Will be adding most recent campaigns on a different PR

<!-- remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm: -->
**_I affirm:_**
- [x] that I agree to Topaz Next's [Limited Contributor License Agreement](https://github.com/topaz-next/topaz/blob/release/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have read the [Contributing Guide](https://github.com/topaz-next/topaz/blob/release/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/topaz-next/topaz/blob/release/CODE_OF_CONDUCT.md)
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

kaincenteno committed:
Login Campaign
Adds logic for adding points if campaign is active.
Makes Greeter Moogles show list of items to trade points for.
Shows messages when zoning/logging in about Campaign.

Only three past campaigns were added, defaulting to December's but it won't run
as that campaign was only meant to run for 23 days.

Will be adding most recent campaigns on a different PR

cocosolos commented on scripts/globals/events/login_campaign.lua:
```diff
@@ -0,0 +1,230 @@
+-- Checks if a Login Campaign is active.
+tpz.events.loginCampaign.isCampaignActive = function()
+    if ENABLE_LOGIN_CAMPAIGN == 1 then
+        local currentLocalTime = os.time(os.date('*t'))
+        local currentUTCTime = os.time(os.date('!*t'))
+        local jstOffset = currentUTCTime - currentLocalTime + 32400
+        local campaignStartDate = os.time({
+            year = loginCampaignYear,
+            month = loginCampaignMonth,
+            day = loginCampaignDay,
+            hour = 0,
+            min = 0,
+            sec = 0
+        }) + jstOffset  -- JST time
```
I don't think this is right. `os.time(os.date('*t'))` will give you the same result as `os.time()`, and `os.date('!*t')` is getting a table for the current date in UTC, then `os.time()` is getting a timestamp of that adjusted date as if you were asking what time UTC it will be when it is that date in _your_ timezone, which is actually UTC+(diff between your time and UTC). So when you subtract `currentLocalTime` you're just setting `currentUTCTime` back to `os.time()`.

Timezone stuff is screwy, but I think this is correct:
```lua
        local local_UTC_offset = os.time() - os.time(os.date('!*t'))
        local JST_UTC_offset = 9 * 60 * 60
        local campaignStartDate = os.time({
            year = loginCampaignYear,
            month = loginCampaignMonth,
            day = loginCampaignDay,
            hour = 0,
            min = 0,
            sec = 0
        }) + local_UTC_offset + JST_UTC_offset
 ```

kaincenteno replied:
Awesome and it makes it more clear, love the use of underscoring here since cammel naming woulda made it messy.
Made the changes, thank you coco <3

kaincenteno committed:
made changes for campaign to follow JST time.
Thank you cocosolos for the suggestion, love the underscoring to make the variable names clear.

TeoTwawki commented on scripts/globals/player.lua:
```diff
@@ -191,6 +192,9 @@ tpz.player.onGameIn = function(player, firstLogin, zoning)
     player:setLocalVar("ZoneInTime", os.time())
+
+    -- Login Campaign rewards points once daily
+    player:timer(2000, function (player) tpz.events.loginCampaign.onGameIn(player) end) -- Hacky recommendation by Teo.
```
I wish I remembered that convo coz I know there are caveats (plural) to any use of timer() but thsi is your only option I think

kaincenteno replied:
yes this was discussed on the former repo PR. Until the zoneIn, onZoneIn issues get resolved this would be a hacky way to get the points =)

claywar replied:
roeutils has an onCharLoad function that is called, maybe something similar could be applied for this?  Also, there's the possibility of joining the two into something more generic.

kaincenteno replied:
Is roeutils a lua util? (i couldnt find it) or is it use in cpp?

zach2good replied:
We should probably add a binding for this in core. `OnGameInLate`, or similar

claywar commented on scripts/globals/events/login_campaign.lua:
```diff
@@ -0,0 +1,229 @@
+tpz.events = tpz.events or {}
+tpz.events.loginCampaign = tpz.events.loginCampaign or {}
+
+-- Change vars below to modify settings for current login campaign
```
Do we have a history of data that could be applied?  I think this would be an excellent opportunity to try to figure out that pattern, and perhaps table it.

kaincenteno replied:
@claywar  I wish there was a pattern, here's a link to all of the login campaigns:
https://www.bg-wiki.com/ffxi/Repeat_Login_Campaign/Past_Login_Campaign_Rewards#Repeat_Login_Campaign_No._1

claywar commented on scripts/globals/events/login_campaign.lua:
```diff
@@ -0,0 +1,229 @@
+            )
+        end
+    end
+end
```
Newline needed for this file

kaincenteno replied:
Added, Thank you =)

claywar commented on scripts/globals/events/login_campaign_data.lua:
```diff
@@ -0,0 +1,463 @@
+-- other lua scripts, which should instead require() login_campaign.lua directly.
+
+-- items table can only hold up to 20 items per option.
+tpz.events.loginCampaign.prizes =
```
Do these IDs shift monthly, or are they part of a larger set determined by the client?

kaincenteno replied:
Yes the prizes shift monthly. Here's a link to the monthly prices.
https://www.bg-wiki.com/ffxi/Repeat_Login_Campaign/Past_Login_Campaign_Rewards#Repeat_Login_Campaign_No._1

claywar replied:
I'm more thinking about the actual vendor options (or mask).  With RoE, there's rotating objectives, but the client just moves to a different mask/value.  For the items that change, is the selection the same?

kaincenteno replied:
the mask changes when you buy an item and then go to a different price. Everytime that you buy an item the mask changes, this is already being handled in login_campaign =)

kaincenteno committed:
Missing newline added. Thanks Claywar =)

claywar commented on scripts/globals/events/login_campaign_data.lua:
```diff
@@ -0,0 +1,463 @@
+                },
+            },
+
+            [5] =
```
These options can be simplified if the value you're passing in is `<<2` prior.  1..25 simplifies to 0..6:
![image](https://user-images.githubusercontent.com/12447174/110208978-06b9e500-7e58-11eb-9dc9-f792ef15a5de.png)

zach2good replied:
^

zach2good commented on scripts/zones/Windurst_Walls/npcs/Greeter_Moogle.lua:
```diff
@@ -0,0 +1,25 @@
+-----------------------------------
+-- Area: Windurst Walls (239)
+--  NPC: Greeter Moogle
+-- !pos NEED POS
```
If these greeter moogles have been capped, you can just grab their pos xyz from the npc sql

zach2good commented on scripts/globals/events/login_campaign_data.lua:
```diff
@@ -0,0 +1,463 @@
+-- This file is called by login_campaign.lua and require()'s no file, it should not be require()'d by any
```
I think its fine to keep this file the size it is. I don't think we need to have a file that holds 3+ years of historic data.

I'd rather have one example entry with lots of options laid out so end-users can copy/paste it and modify it to their liking.

zach2good replied:
Also, the last thing we want is a PR every month adding a new entry to this file, invalidating everything that has come before. We're interested in emulating current retail; so at best new entries will only be "valid" for a couple of weeks before fading into obscurity
